src/miner_gateway_ecc_worker.erl

```diff
@@ -0,0 +1,148 @@
+%%%-------------------------------------------------------------------
+%% @doc miner gateway ecc worker
+%% Interface to the rust-based gateway service over grpc
+%% @end
+%%%-------------------------------------------------------------------
+-module(miner_gateway_ecc_worker).
+
+-behaviour(gen_server).
+
+-export([pubkey/0,
+         ecdh/1,
+         sign/1,
+         reconnect/0]).
+
+-export([start_link/1,
+         init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2]).
+
+-record(state, {
+                 connection :: pid(),
+                 host="localhost" :: string(),
+                 port=4468 :: integer(),
+                 transport=tcp :: tcp | ssl
+               }).
+
+-define(RETRY_WAIT, 10).
+-define(MAX_RETRIES, 10).
+%% Make the call timeout quite long since this worker has to process
+%% all signing requests for the system
+-define(CALL_TIMEOUT, (?MAX_RETRIES * ?RETRY_WAIT) * 10).
+
+%% Retrieve the public key for the ecc chip from the rust gateway and
+%% deserialize it from the binary returned
+-spec pubkey() -> {ok, libp2p_crypto:pubkey()} | {error, term()}.
+pubkey() ->
+    gen_server:call(?MODULE, pubkey, ?CALL_TIMEOUT).
+
+%% Pass a binary to the rust gateway for signing and returned the signed binary
+-spec sign(binary()) -> {ok, Signature::binary() | {error, term()}}.
+sign(Binary) ->
+    gen_server:call(?MODULE, {sign, Binary}, ?CALL_TIMEOUT).
+
+%% Pass an ecc public key to the rust gateway and return a point on the
+%% eliptic curve as a binary
+-spec ecdh(libp2p_crypto:pubkey()) -> {ok, Preseed::binary()} | {error, term()}.
+ecdh({ecc_compact, _Bin} = PubKey) ->
+    gen_server:call(?MODULE, {ecdh, PubKey}, ?CALL_TIMEOUT).
+
+%% Trigger a reconnect of the grpc_client connection
+-spec reconnect() -> ok.
+reconnect() ->
+    gen_server:cast(?MODULE, reconnect).
+
+start_link(Options) when is_list(Options) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Options], []).
+
+init([Options]) ->
+    Transport = proplists:get_value(transport, Options, tcp),
+    Host = proplists:get_value(host, Options, "localhost"),
+    Port = proplists:get_value(port, Options, 4468),
+    {ok, Connection} = grpc_connect(Transport, Host, Port),
```

Comment:
maybe push this out of the init.  grpc_connect has reconnect support with 10 seconds between retries.  Handling this in init could potentially hang your supervisor startup sequence whilst it waits for grpc connect to complete

Reply:
since this is now our only point of access to the pubkey which we need to pass to the downstream supervisors (blockchain_sup in critical path at least) and also those signing and ecdh functions, i would think we would _want_ to explicitly hang in the supervisor init block here in order to ensure the other supervisors have the info they need. if we somehow didn't connect immediately and had to wait 10 seconds in order to retrieve the pubkey i think the miner would panic

Reply:
k, i hadnt noted that and makes sense.  I will follow through the sup sequence again